plat/kvm/x86/multiboot2.c

```diff
@@ -0,0 +1,180 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/essentials.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/types.h>
+#include <uk/arch/paging.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/plat/common/lcpu.h>
+#include <uk/plat/common/memory.h>
+#include <uk/plat/common/sections.h>
+#include <uk/reloc.h>
+#include <kvm-x86/multiboot2.h>
+
+#include <errno.h>
+#include <string.h>
+
+#define multiboot2_crash(msg, rc)	ukplat_crash()
+
+/* Round up 'x' to the nearest multiple of 'MULTIBOOT_TAG_ALIGN' */
+#define ROUND_UP_TO_ALIGNMENT(x) \
+	(((x) + (MULTIBOOT_TAG_ALIGN - 1)) & ~(MULTIBOOT_TAG_ALIGN - 1))
+
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static inline void mrd_insert(struct ukplat_bootinfo *bi,
+			      const struct ukplat_memregion_desc *mrd)
+{
+	int rc;
+
+	if (unlikely(mrd->len == 0))
+		return;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, mrd);
+	if (unlikely(rc < 0))
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+/**
+ * Multiboot entry point called after lcpu initialization. We enter with the
+ * 1:1 boot page table set. Physical and virtual addresses thus match for all
+ * regions in the mapped range.
+ */
+void multiboot2_entry(struct lcpu *lcpu, unsigned long addr)
+{
+	struct ukplat_bootinfo *bi;
+	struct ukplat_memregion_desc mrd = {0};
+	multiboot_memory_map_t *m;
+	__u64 cmdline_str = {0};
+	__sz i, cmdline_len, mmap_entry_cnt;
+	__paddr_t start, end;
+	int rc;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
+		multiboot2_crash("Incompatible or corrupted bootinfo", -EINVAL);
+
+	/* We have to call this here as the very early do_uk_reloc32 relocator
+	 * does not also relocate the UKPLAT_MEMRT_KERNEL mrd's like its C
+	 * equivalent, do_uk_reloc, does.
+	 */
+	do_uk_reloc_kmrds(0, 0);
+
+	/* Ensure that the memory map contains the legacy high mem area */
+	rc = ukplat_memregion_list_insert_legacy_hi_mem(&bi->mrds);
+	if (unlikely(rc))
+		multiboot2_crash("Could not insert legacy memory region", rc);
+
+	/**
+	 * Initialize the first tag address, skipping total_size(4) and
+	 * reserved (4) fields of the boot information structure
+	 */
+	struct multiboot_tag *tag = (struct multiboot_tag *)(addr + 8);
+
+	/* Parse Multiboot2 tags */
+	for (; tag->type != MULTIBOOT_TAG_TYPE_END;
+	     tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					+ ROUND_UP_TO_ALIGNMENT(tag->size))) {
```

Comment:
Yeah this will look super ugly as a for loop given how long these definitions are... maybe let's do it as a while loop instead? :)